### DISCUSSION TITLE
Hardware debouncing for Encoder

### DISCUSSION THREAD

**Discussion by luisfcorreia** [probonopd/MiniDexed#178] [Ideas]
We could use this simple circuit for debouncing the 3 switches from the encoder.

This is part of Norns Shield hardware (https://github.com/monome/norns-shield)

![From Norns Sheild](https://user-images.githubusercontent.com/191885/165369625-9be9cf0d-8714-49eb-842b-ac2ea0fb4892.png)

**Comment by probonopd** [comment]
What would the effect of this be, in laymans terms? (Isn't bouncing resulting in too many pulses? We seem to get/use too few, if anything.)
If possible, I'd like to do without adding more hardware besides a display module and an encoder module.

**Reply by probonopd** [reply]
I only had 2 1nF caps at hand. Adding those to the A and B lines of the encoder already makes a big difference indeed! Far fewer missed steps. Will order 10nF ones.

**Comment by luisfcorreia** [comment]
instead of wasting CPU cycles trying to wait for the pins to stabilize, this circuit does it externally

it's very low cost, just works

(most of the times, you can just add the capacitors C35, C36 and C44)

**Comment by bmillier** [comment]
I found that just adding 10 nF caps from the rotary encoder pins to ground works fine. I have some good encoders, but the only encoders I have with the P.Button switch are cheap little Bourns ones, and without the caps they are quite erratic- miss pulses or skip over 2 detents, etc.

**Reply by probonopd** [reply]
Hi @[bmillier](https://github.com/bmillier), are you saying that using the caps you get one step in the software reliably when you turn the encoder by one physical step?

**Comment by bmillier** [comment]
Hi probono (Pro Bono - are you a lawyer?) 
   Yes, it is quite acceptable with just the caps. As I said, the tiny Bourns R.E.s that I have on hand have quite a contact bounce. I only use them in my projects when I need the P.B. switch option and then I de-bounce them in MCU firmware.
 Luiscorreia's circuit is better- it has dedicated low value pull-ups which are better than the ~100k ones in the Raspi's GPIO internal pull-up option. But, its quite a few components unless you put them on a board- my 2 caps to ground just mount directly from RE A,B terminals to the COM terminals.

**Comment by craigyjp** [comment]
Added just the 3x 10nF caps today across the encoder and it responds perfectly now.

**Reply by luisfcorreia** [reply]
I added 100nF and it _also_ works great!
